cluster-autoscaler/cloudprovider/kwok/kwok_helpers.go

```diff
@@ -266,13 +271,5 @@ func getNGName(no *apiv1.Node, kc *KwokProviderConfig) string {
 		klog.Fatal("grouping criteria for nodes is not set (expected: 'annotation' or 'label')")
 	}
 
-	if ngName == "" {
-		klog.Fatalf("%s '%s' for node '%s' not present in the manifest",
-			kc.status.groupNodesBy, kc.status.key,
-			no.GetName())
-	}
-
-	ngName = fmt.Sprintf("%s-%v", ngName, time.Now().Unix())
```

Comment:
I think it might be better to keep the node group name unchanged, especially in cases where nodes still remain in the cluster and CA is restarted.

Reply:
Line 275 is clearly a bug. The original intention was to ensure `targetSize` of a nodegroup correctly reflects the number of nodes with matching annotation/label. 

If we already have nodes in the cluster with matching nodegroup annotations/labels and we remove `time.Now().Unix()` suffix, when creating nodegroups (imagine the CA pod got restarted) the target size won't accurately reflect the actual nodegroup size because there are more nodes matching nodegroup label in the cluster now. 

One solution can be to implement [`Refresh`](https://github.com/kubernetes/autoscaler/blob/df6784bfafa7e94011ddae1f578fcb334566f88c/cluster-autoscaler/cloudprovider/kwok/kwok_provider.go#L127) and update the nodegroup target size based on the actual matching nodes in the cluster.

Reply:
Let me know what you think.

Reply:
The Kwok provider will calculate the target sizes during startup, and the Cluster Autoscaler will scale these nodegroups to the appropriate size. I think it's not necessary to calculate the target size in the `Refresh` function, just leave this work to CA may be a better choice.

Reply:
kwok provider will clean all the fake node when CA quit:

```golang
// Cleanup cleans up all resources before the cloud provider is removed
func (kwok *KwokCloudProvider) Cleanup() error {
	for _, ng := range kwok.nodeGroups {
		nodeNames, err := ng.getNodeNamesForNodeGroup()
		if err != nil {
			return fmt.Errorf("error cleaning up: %v", err)
		}

		for _, node := range nodeNames {
			err := kwok.kubeClient.CoreV1().Nodes().Delete(context.Background(), node, v1.DeleteOptions{})
			if err != nil {
				klog.Errorf("error cleaning up kwok provider nodes '%v'", node)
			}
		}
	}

	return nil
}
```

Reply:
True. I was considering a case when `Cleanup` doesn't clean all the nodes (it doesn't work correctly for whatever reason)

> The Kwok provider will calculate the target sizes during startup, and the Cluster Autoscaler will scale these nodegroups to the appropriate size. I think it's not necessary to calculate the target size in the Refresh function, just leave this work to CA may be a better choice.

I think we might have to try this out to confirm if CA can handle such a situation. If you can try it out as a part of this PR, great. If not, we can take care of it in another issue.

Reply:
Hi @vadasambar, I have implemented the Refresh() function. Please take a look.